File: bingo/views/report.py
from django.contrib.auth.decorators import login_required
from django.http import JsonResponse
from django.utils.timezone import now
from decimal import Decimal

from bingo.models import BingoTransaction, BingoUser

@login_required
def fetch_reports(request):
    user = request.user
    start_date = request.GET.get("start_date")
    # end_date = request.GET.get("end_date")

    transactions = BingoTransaction.objects.filter(daily_record__user__owner=user)
    if not start_date:
        transactions = transactions.filter(daily_record__date=now().date())
    if start_date:
        transactions = transactions.filter(daily_record__date=start_date)
    # if end_date:
    #     transactions = transactions.filter(daily_record__date__lte=end_date)

    bingo = BingoUser.objects.get(owner=request.user)
    # Calculate totals
    total_transactions = transactions.count()
    total_balance = Decimal(bingo.balance)
    total_winning = sum(transaction.total_won for transaction in transactions)
    total_cut = sum(transaction.cut for transaction in transactions)

    data = [
        {
            "date": transaction.daily_record.date.strftime("%Y-%m-%d"),
            "time": transaction.time.strftime("%H:%M:%S"),
            "bet": float(transaction.bet),
            "player_number": transaction.player_number,
            "total_won": float(transaction.total_won),
            "cut": float(transaction.cut),
            "won": float(transaction.won),
            "call_number": transaction.call_number,
            "winners":transaction.winners if transaction.winners else transaction.submitted_cartella,
            "branch": transaction.daily_record.user.branch,
            "cashier": transaction.daily_record.user.owner.username,
            "balance": float(transaction.daily_record.user.balance),
        }
        for transaction in transactions
    ]
    # print(f'report data: {data}')

    return JsonResponse({
        "data": data,
        "total_transactions": total_transactions,
        "total_balance": total_balance,
        "total_winning": total_winning,
        "total_cut": total_cut,
    })

File: static/bingo_static/js/report.js

$(document).ready(function () {
    // Show the modal
    $("#openDBModal").on("click", function () {
        $("#dateSearchModal").show();
    });

    // Hide the modal
    $("#closeReportModal").on("click", function () {
        $("#dateSearchModal").hide();
    });

    // Handle the date search form submission
    $("#dateSearchForm").on("submit", function (e) {
        e.preventDefault();
        const startDate = $("#startDate").val();
        // const endDate = $("#endDate").val();
        fetchReports(startDate);
        // fetchReports(startDate, endDate);
    });

    function fetchReports(startDate = "", endDate = "") {
        $.ajax({
            url: "/reports/",
            method: "GET",
            data: {
                start_date: startDate,
                // end_date: endDate,
            },
            success: function (response) {
                const tableBody = $("#reportsTable tbody");
                tableBody.empty(); // Clear previous data

                // Update header with totals
                const modalHeader = $(".report-header");
                const totalBalance = parseFloat(response.total_balance) || 0;
                const totalTransactions = response.total_transactions || 0;
                const totalWinning = parseFloat(response.total_winning) || 0;
                const totalCut = parseFloat(response.total_cut) || 0;

                // Update left, middle, and right sections
                modalHeader.find(".report-left").html(`
                    <div class="report-item">
                        <span>Game Transactions:</span>
                        <strong>${totalTransactions}</strong>
                    </div>
                `);

                modalHeader.find(".report-middle h3").html(`
                    <div class="report-item" style="color: red; font-size: 1.4rem">
                        <span>Balance:</span>
                        <strong>${totalBalance.toFixed(2)} Birr</strong>
                    </div>
                `);

                modalHeader.find(".report-right").html(`
                    <div class="report-item" style="color: white;">
                        <span>Winning:</span>
                        <strong>${totalWinning.toFixed(2)}</strong>
                    </div>
                    <div class="report-item" style="color: white;>
                        <span>Cut:</span>
                        <strong>${totalCut.toFixed(2)}</strong>
                    </div>
                `);
                if (response.data.length === 0) {
                    tableBody.append("<tr><td colspan='12'>No data available</td></tr>");
                } else {
                    response.data.forEach((row) => {
                        tableBody.append(`
                      <tr>
                          <td>${row.date}</td>
                          <td>${row.time}</td>
                          <td>${row.bet}</td>
                          <td>${row.player_number}</td>
                          <td>${row.total_won}</td>
                          <td>${row.cut}</td>
                          <td>${row.won}</td>
                          <td>${row.call_number}</td>
                          <td>${row.winners}</td>
                          <td>${row.branch}</td>
                          <td>${row.cashier}</td>
                      </tr>
                  `);
                    });
                }
            },
            error: function () {
                showModalAlert("Failed to fetch reports. Please try again.");
            },
        });
    }
});